docs/Customization/Custom-Core.rst

```diff
@@ -0,0 +1,404 @@
+breakpoints (set by ``nBreakpoints`` in ``CoreParams``). Below is the definition of these types:
+
+.. code-block:: scala
+
+    // Control signal from the external tracer
+    class TraceAux extends Bundle {
+      val enable = Bool()   // Enable trace output
+      val stall = Bool()    // If true, the core should stall
+    }
+    // Check RISC-V Processor Trace spec V1.0 for more information of this interface
+    class TraceCoreInterface (val params: TraceCoreParams) extends Bundle {
+      val group = Vec(params.nGroups, new TraceCoreGroup(params))
+      val priv = UInt(4.W)
+      val tval = UInt(params.xlen.W)
+      val cause = UInt(params.xlen.W)
+    }
+    // Address Breakpoint and watchpoint info (n is the retire width)
+    class BPWatch (val n: Int) extends Bundle() {
+      val valid = Vec(n, Bool())    // Valid bit of the output
+      val rvalid = Vec(n, Bool())   // Break on read
+      val wvalid = Vec(n, Bool())   // Break on write
+      val ivalid = Vec(n, Bool())   // Break on execute
+      val action = UInt(3.W)        // Exception code (3 usually) 
+    }
+
+Implementation Class
+--------------------
+
+The implementation class is of the following form:
+
+.. code-block:: scala
+
+    class MyTileModuleImp(outer: MyTile) extends BaseTileModuleImp(outer){
+      // annotate the parameters
+      Annotated.params(this, outer.tileParams)
+
+      // TODO: Create the top module of the core and connect it with the ports in "outer"
+    }
+
+In the body of this class, you can look up any parameters by calling ``p({key})``, where ``{key}`` is the config key of 
+the value you want to look up. For a list of available keys, see the appendix below.
+
+If you create an AXI4 node (or equivalents), you will need to connect them to your core. You can connect a port like this:
+
+.. code-block:: scala
+
+    outer.myAXI4Node.out foreach { case (out, edgeOut) =>
+      // Connect your module IO port to "out"
+      // The type of "out" here is AXI4Bundle, which is defined in generators/rocket-chip/src/main/scala/amba/axi4/Bundles.scala
+      // Please refer to this file for the definition of the ports.
+      // If you are using APB, check APBBundle in generators/rocket-chip/src/main/scala/amba/apb/Bundles.scala
+      // If you are using AHB, check AHBSlaveBundle or AHBMasterBundle in generators/rocket-chip/src/main/scala/amba/ahb/Bundles.scala
+      // (choose one depends on the type of AHB node you create)
+      // If you are using AXIS, check AXISBundle and AXISBundleBits in generators/rocket-chip/src/main/scala/amba/axis/Bundles.scala
+    }
+
+Integrate the Core
+------------------
+
+To use your core in a set of config, you would need a config fragment that would create a ``TileParams`` object of your core in
+the current config. An example of such config will be like this:
+
+.. code-block:: scala
+
+    class WithNMyCores(n: Int) extends Config((site, here, up) => {
```

Comment:
This should have the `hartidOffset` field.

Reply:
Done.